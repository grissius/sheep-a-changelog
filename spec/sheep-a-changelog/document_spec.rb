require 'sheep-a-changelog/node'
require 'sheep-a-changelog/document'

changelog = File.read(File.expand_path('examples/keepachangelog.md', __dir__))

RSpec.describe SheepAChangelog::Document do
  subject(:doc) { SheepAChangelog::Document.new(changelog.split("\n")) }

  it 'is same as top-level node' do
    expect(doc.build_tree).to eq(SheepAChangelog::Node.new(changelog.split("\n"), :empty, 0).build_tree)
  end

  it 'version_root' do
    expect(doc.version_root.nodes.map(&:title)).to match_snapshot
  end

  it 'first_version' do
    expect(doc.first_version).to match(/Unreleased/)
  end

  it 'diff_prefix' do
    expect(doc.diff_prefix).to match('https://github.com/olivierlacan/keep-a-changelog/compare/')
  end

  it 'rename_version' do
    new_version = 'foo_bar'
    doc.rename_version('[Unreleased]', new_version)
    expect(doc.first_version).to match(new_version)
    expect(doc.build_tree).to match_snapshot
  end

  it 'add_anchor' do
    new_version = 'foo_bar'
    doc.add_anchor('LABEL', 'vFROM', 'vTO')
    expect(doc.anchors).to match_snapshot
  end
end
