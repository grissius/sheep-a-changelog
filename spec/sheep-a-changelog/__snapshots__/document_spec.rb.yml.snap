---
"./spec/sheep-a-changelog/document_spec.rb[1:2]":
- "[Unreleased]"
- "[1.0.0] - 2017-06-20"
- "[0.3.0] - 2015-12-03"
- "[0.2.0] - 2015-10-06"
- "[0.1.0] - 2015-10-06"
- "[0.0.8] - 2015-02-17"
- "[0.0.7] - 2015-02-16"
- "[0.0.6] - 2014-12-12"
- "[0.0.5] - 2014-08-09"
- "[0.0.4] - 2014-08-09"
- "[0.0.3] - 2014-08-09"
- "[0.0.2] - 2014-07-10"
- 0.0.1 - 2014-05-31
"./spec/sheep-a-changelog/document_spec.rb[1:5]":
  :title: :empty
  :lines: []
  :anchors:
  - :v: Unreleased
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v1.0.0...HEAD
  - :v: 1.0.0
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.3.0...v1.0.0
  - :v: 0.3.0
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.2.0...v0.3.0
  - :v: 0.2.0
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.1.0...v0.2.0
  - :v: 0.1.0
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.8...v0.1.0
  - :v: 0.0.8
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.7...v0.0.8
  - :v: 0.0.7
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.6...v0.0.7
  - :v: 0.0.6
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.5...v0.0.6
  - :v: 0.0.5
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.4...v0.0.5
  - :v: 0.0.4
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.3...v0.0.4
  - :v: 0.0.3
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.2...v0.0.3
  - :v: 0.0.2
    :url: https://github.com/olivierlacan/keep-a-changelog/compare/v0.0.1...v0.0.2
  :nodes:
  - :title: Changelog
    :lines:
    - All notable changes to this project will be documented in this file.
    - ''
    - The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
    - and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
    - ''
    :anchors: []
    :nodes:
    - :title: foo_bar
      :lines: []
      :anchors: []
      :nodes:
      - :title: Changed
        :lines:
        - "- Update and improvement of Polish translation from [@m-aciek](https://github.com/m-aciek)."
        - ''
        :anchors: []
        :nodes: []
    - :title: "[1.0.0] - 2017-06-20"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - "- New visual identity by [@tylerfortune8](https://github.com/tylerfortune8)."
        - "- Version navigation."
        - "- Links to latest released version in previous versions."
        - '- "Why keep a changelog?" section.'
        - '- "Who needs a changelog?" section.'
        - '- "How do I make a changelog?" section.'
        - '- "Frequently Asked Questions" section.'
        - '- New "Guiding Principles" sub-section to "How do I make a changelog?".'
        - "- Simplified and Traditional Chinese translations from [@tianshuo](https://github.com/tianshuo)."
        - "- German translation from [@mpbzh](https://github.com/mpbzh) & [@Art4](https://github.com/Art4)."
        - "- Italian translation from [@azkidenz](https://github.com/azkidenz)."
        - "- Swedish translation from [@magol](https://github.com/magol)."
        - "- Turkish translation from [@karalamalar](https://github.com/karalamalar)."
        - "- French translation from [@zapashcanon](https://github.com/zapashcanon)."
        - "- Brazilian Portugese translation from [@Webysther](https://github.com/Webysther)."
        - "- Polish translation from [@amielucha](https://github.com/amielucha)."
        - "- Russian translation from [@aishek](https://github.com/aishek)."
        - "- Czech translation from [@h4vry](https://github.com/h4vry)."
        - "- Slovak translation from [@jkostolansky](https://github.com/jkostolansky)."
        - "- Korean translation from [@pierceh89](https://github.com/pierceh89)."
        - "- Croatian translation from [@porx](https://github.com/porx)."
        - "- Persian translation from [@Hameds](https://github.com/Hameds)."
        - ''
        :anchors: []
        :nodes: []
      - :title: Changed
        :lines:
        - '- Start using "changelog" over "change log" since it''s the common usage.'
        - "- Start versioning based on the current English version at 0.3.0 to help"
        - translation authors keep things up-to-date.
        - '- Rewrite "What makes unicorns cry?" section.'
        - '- Rewrite "Ignoring Deprecations" sub-section to clarify the ideal'
        - "  scenario."
        - '- Improve "Commit log diffs" sub-section to further argument against'
        - "  them."
        - '- Merge "Why can’t people just use a git log diff?" with "Commit log'
        - '  diffs"'
        - "- Fix typos in Simplified Chinese and Traditional Chinese translations."
        - "- Fix typos in Brazilian Portuguese translation."
        - "- Fix typos in Turkish translation."
        - "- Fix typos in Czech translation."
        - "- Fix typos in Swedish translation."
        - "- Improve phrasing in French translation."
        - "- Fix phrasing and spelling in German translation."
        - ''
        :anchors: []
        :nodes: []
      - :title: Removed
        :lines:
        - '- Section about "changelog" vs "CHANGELOG".'
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.3.0] - 2015-12-03"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - "- RU translation from [@aishek](https://github.com/aishek)."
        - "- pt-BR translation from [@tallesl](https://github.com/tallesl)."
        - "- es-ES translation from [@ZeliosAriex](https://github.com/ZeliosAriex)."
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.2.0] - 2015-10-06"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Changed
        :lines:
        - '- Remove exclusionary mentions of "open source" since this project can'
        - benefit both "open" and "closed" source projects equally.
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.1.0] - 2015-10-06"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - '- Answer "Should you ever rewrite a change log?".'
        - ''
        :anchors: []
        :nodes: []
      - :title: Changed
        :lines:
        - "- Improve argument against commit logs."
        - "- Start following [SemVer](https://semver.org) properly."
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.0.8] - 2015-02-17"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Changed
        :lines:
        - "- Update year to match in every README example."
        - "- Reluctantly stop making fun of Brits only, since most of the world"
        - "  writes dates in a strange way."
        - ''
        :anchors: []
        :nodes: []
      - :title: Fixed
        :lines:
        - "- Fix typos in recent README changes."
        - "- Update outdated unreleased diff link."
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.0.7] - 2015-02-16"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - "- Link, and make it obvious that date format is ISO 8601."
        - ''
        :anchors: []
        :nodes: []
      - :title: Changed
        :lines:
        - '- Clarified the section on "Is there a standard change log format?".'
        - ''
        :anchors: []
        :nodes: []
      - :title: Fixed
        :lines:
        - "- Fix Markdown links to tag comparison URL with footnote-style links."
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.0.6] - 2014-12-12"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - '- README section on "yanked" releases.'
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.0.5] - 2014-08-09"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - "- Markdown links to version tags on release headings."
        - "- Unreleased section to gather unreleased changes and encourage note"
        - keeping prior to releases.
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.0.4] - 2014-08-09"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - '- Better explanation of the difference between the file ("CHANGELOG")'
        - and its function "the change log".
        - ''
        :anchors: []
        :nodes: []
      - :title: Changed
        :lines:
        - '- Refer to a "change log" instead of a "CHANGELOG" throughout the site'
        - to differentiate between the file and the purpose of the file — the
        - logging of changes.
        - ''
        :anchors: []
        :nodes: []
      - :title: Removed
        :lines:
        - "- Remove empty sections from CHANGELOG, they occupy too much space and"
        - create too much noise in the file. People will have to assume that the
        - missing sections were intentionally left out because they contained no
        - notable changes.
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.0.3] - 2014-08-09"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - '- "Why should I care?" section mentioning The Changelog podcast.'
        - ''
        :anchors: []
        :nodes: []
    - :title: "[0.0.2] - 2014-07-10"
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - "- Explanation of the recommended reverse chronological release ordering."
        - ''
        :anchors: []
        :nodes: []
    - :title: 0.0.1 - 2014-05-31
      :lines: []
      :anchors: []
      :nodes:
      - :title: Added
        :lines:
        - "- This CHANGELOG file to hopefully serve as an evolving example of a"
        - "  standardized open source project CHANGELOG."
        - "- CNAME file to enable GitHub Pages custom domain"
        - "- README now contains answers to common questions about CHANGELOGs"
        - "- Good examples and basic guidelines, including proper date formatting."
        - '- Counter-examples: "What makes unicorns cry?"'
        - ''
        :anchors: []
        :nodes: []
